client.go

```diff
@@ -0,0 +1,289 @@
+	}
+
+	// send the request
+	resp, err := treq.client.cfg.httpClient.Do(req)
+	if err != nil {
+		// errors from Do() may be permanent or not, it is not possible to
+		// determine precisely
+		return // allow retry
+	}
+
+	// handle response
+	if resp.StatusCode >= 400 {
+		err = handleErrorResponse(resp, logger)
+		logFullRequestError(logger, treq, reqBody, err)
+		return
+	}
+
+	tresp, err = handleSuccessResponse[TRESP](resp, logger)
+	if err != nil {
+		logFullRequestError(logger, treq, reqBody, err)
+		return
+	}
+
+	if treq.client.cfg.logSuccess {
+		logFullHTTPRequestSuccess(logger, treq, reqBody, tresp)
+	}
+
+	return tresp, err
+}
+
+func handleSuccessResponse[TRESP commonResponseSetter](httpResp *http.Response, logger *logs.Logger) (
+	*response[TRESP],
+	error,
+) {
+	var ret response[TRESP]
+
+	ret.code = httpResp.StatusCode
+	ret.headers = httpResp.Header
+
+	var err error
+	ret.rawBody, err = io.ReadAll(httpResp.Body)
+	if err != nil {
+		// allow retry
+		return nil, errors.Join(err, HTTPError{
+			Code:    httpResp.StatusCode,
+			RawBody: ret.rawBody,
+			Headers: ret.headers,
+		})
+	}
+
+	err = json.Unmarshal(ret.rawBody, &ret.body)
+	if err != nil {
+		// allow retry
+		return nil, errors.Join(err, HTTPError{
+			Code:    httpResp.StatusCode,
+			RawBody: ret.rawBody,
+			Headers: ret.headers,
+		})
+	}
+
+	return &ret, nil
+}
+
+func handleErrorResponse(resp *http.Response, logger *logs.Logger) error {
+	// the error response must always support errors.As(err, HTTPError)
+	httpErr := HTTPError{
+		Code:    resp.StatusCode,
+		Headers: resp.Header,
+	}
+
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("CloudFlare returned an error, but failed to read the error body: %v; %w", err, httpErr) // allow retry
+	}
+
+	httpErr.RawBody = respBody
+
+	// try to parse the CloudFlare error objects
+	mediaType, _, err := mime.ParseMediaType(resp.Header.Get("content-type"))
+	if err != nil {
+		return retry.PermanentError{Cause: fmt.Errorf("CloudFlare returned an error, and the content-type of the response is unsupported %q (%v); %w",
+			resp.Header.Get("content-type"), err, httpErr)}
+	}
+
+	if mediaType != "application/json" {
+		return retry.PermanentError{Cause: fmt.Errorf("CloudFlare returned an error, and the content-type of the response is unsupported %q\n%w",
+			resp.Header.Get("content-type"), httpErr)}
+	}
+
+	var cfcommon cfResponseCommon
+	err = json.Unmarshal(respBody, &cfcommon)
+	if err != nil {
+		return retry.PermanentError{Cause: fmt.Errorf("CloudFlare returned an error, but failed to read the error body: %v; %w", err, httpErr)}
+	}
+
+	ret := CloudFlareError{
+		cfResponseCommon: cfcommon,
+		HTTPError:        httpErr,
+	}
+
+	if httpErr.IsPermanent() {
+		return retry.PermanentError{Cause: ret}
+	}
+
+	return ret
+}
+
+func logFullRequestError[T any](logger *logs.Logger, treq *request[T], reqBody []byte, err error) {
+	logger.D(func(log logs.DebugFn) {
+		msg := &bytes.Buffer{}
```

Comment:
optional:
use a strings.Builder instead?

Reply:
`strings.Builder` only gives `WriteString`, so I still have to use `fmt.Fprintf` or `fmt.Sprintf`. I prefer using only `bytes.Buffer` and using `fmt.Fprint*` for everything.

Reply:
I'm not sure if I understand your comment.
You can also use fmt.Fprintln with a string builder.

This should be the only necessary change:
```suggestion
		msg := &strings.Builder{}
```

(might not be needed anymore when `httputil.DumpResponseOut` would also be used)

Reply:
Changed to use `string.Builder`. A builder/buffer is still necessary.